File: Week2/node/main.js
const { ask, closeInput } = require("./utils/input");
const { parseNumberList } = require("./utils/parse");
const { printMainMenu } = require("./menu/menuPrinter");
const { runSearchFlow } = require("./flows/searchFlow");
const { runSortFlow } = require("./flows/sortFlow");

async function getArrayFromUser() {
  while (true) {
    const raw = await ask(
      "\nEnter array numbers (space or comma separated), e.g. 10 3 25 7 1: "
    );

    const parsed = parseNumberList(raw);
    if (parsed.ok) return parsed.value;

    console.log(`Error: ${parsed.error} Try again.`);
  }
}

async function app() {
  while (true) {
    console.log("\n=== Welcome ===");
    console.log("1. Start");
    console.log("0. Exit");

    const startChoice = await ask("Select an option: ");

    if (startChoice === "0") {
      console.log("Goodbye!");
      return;
    }

    if (startChoice === "1") {
      break;
    }

    console.log("Invalid option. Try again.");
  }

  const arr = await getArrayFromUser();

  while (true) {
    printMainMenu();
    const action = await ask("Select an action (1-2) or 0 to Exit: ");

    if (action === "0") {
      console.log("Goodbye!");
      return;
    }

    if (action === "1") {
      await runSearchFlow(arr);
      continue;
    }

    if (action === "2") {
      await runSortFlow(arr);
      continue;
    }

    console.log("Invalid action.");
  }
}


(async () => {
  try {
    await app();
  } catch (err) {
    console.error("Something went wrong:", err?.message ?? err);
    process.exitCode = 1;
  } finally {
    closeInput();
  }
})();

File: Week2/python/main.py
from menu.menu_printer import print_main_menu
from flows.search_flow import run_search_flow
from flows.sort_flow import run_sort_flow
from utils.parse import parse_number_list


def get_array_from_user():
    while True:
        raw = input("\nEnter array numbers (space or comma separated), e.g. 10 3 25 7 1: ")
        parsed = parse_number_list(raw)

        if parsed["ok"]:
            return parsed["value"]

        print(f"Error: {parsed['error']} Try again.")


def show_welcome_menu():
    print("\n=== Welcome (Python) ===")
    print("1. Start")
    print("0. Exit")


def main():
    try:
        # ✅ Welcome step (Start/Exit) BEFORE asking for the array
        while True:
            show_welcome_menu()
            start_choice = input("Select an option: ").strip()

            if start_choice == "0":
                print("Goodbye!")
                return

            if start_choice == "1":
                break

            print("Invalid option. Try again.")

        # Only ask for array after user chooses Start
        arr = get_array_from_user()

        while True:
            print_main_menu()
            action = input("Select an action (1-2) or 0 to Exit: ").strip()

            if action == "0":
                print("Goodbye!")
                return

            if action == "1":
                run_search_flow(arr)
                continue

            if action == "2":
                run_sort_flow(arr)
                continue

            print("Invalid action. Please choose 1, 2, or 0.")

    except KeyboardInterrupt:
        print("\nExiting...")


if __name__ == "__main__":
    main()
